### ISSUE TITLE
Improve Annotations vs Glyphs for 3.0

### ISSUE THREAD

**Issue by bryevdv** [bokeh/bokeh#9955]
There are duplications, inconsistencies and warts around annotations. 

Some context: The original motivation for splitting off annotations like `BoxAnnotation` from say, a `Rect` glyph  is that annotations may be positioned via screen coordinates, which makes re-using the spatial index hit-testing cumbersome/impossible. 

Some goals:

* Adopt consistent naming conventions for single/multiple. E.g. we have `Label` (single) and `LabelSet` (plural) for labels. But Arrow i implicitly vectorized with a source. 

* Allow annotations to be added to legends (this has come up several times recently)

* Allow annotations to participate in hit testing -- but this is probably best exposed as a new event for annotations, rather than selections on data sources  

* Some annotation specific improvements: 
  - make it easy to attach labels to other things
  - make arrows simpler/easier to use (also easy to "attach" to things)

* Try to find opportunities to reduce code duplication between annoations and similar glyphs

cc @bokeh/dev

**Comment by p-himik** [comment]
I haven't given it much thought yet, but I would definitely like to see the line separating annotations and glyphs as non-intrusive as possible. Ideally, non-existent at all. E.g. it would be nice to be able to specify screen coordinates for glyphs as well - e.g. as a way to create complex annotations. But I have no idea how feasible it is.

**Comment by jbednar** [comment]
I'd love to see any improvements in this area. I too often find that the boundary between annotations and glyphs is or should be fluid.  It would be great if all visible objects could be specified in either screen or data coordinates, as needed, if all annotations had a vectorized version where they act more like data, etc.  I agree that in the extremes there are differences between annotations and glyphs, but very often one starts at one of those extremes and then wants to do some glyph-like things for annotations or some annotation-like things for glyphs, so the differences between those two categories I think are best understood as differing sets of default values rather than non-overlapping sets of functionality.

**Comment by bryevdv** [comment]
So, FYI I agre things should be made less obtrusive, think there are reasonable reasons to maintain some separation:

* annotation should generally go "on top" so the separation affords a place to specify an appropriate default render level for annotations that is different from glyphs

* rbush spatial indexing uses data coords. It has to, otherwise a complete-reindex would be required on  e.g. every single pan movement.  But this makes hit testing things that can have fixed screen coordinates (or worse--a mix of screen and data coordinates) very, very difficult. 

  What  would like is for annotations to support hit tests, just not to use rbush for it like glyphs do (there are usually only a relatively small number of annotations drawn, hit testing can just be brute-forced in screen space)

**Comment by jbednar** [comment]
Definitely, that's all reasonable. Docs should have an annotations category, default render level for annotations should be different, etc.  Basically, there is a story we can tell about the two types of visible object, based on their expected or intended usage, and the default values of various settings can and should be different for these two categories of object that we present separately.  And for technical reasons there can be limitations in the fluidity between the two categories, for sound technical reasons (as for hit testing).  But I think that thinking of the two categories as each having some "pure" examples in them that are at the extreme points of a continuum, while many of them also can be fluildly interpreted as being in either camp or as occupying a wide range of that continuum, is helpful...

**Comment by p-himik** [comment]
https://discourse.bokeh.org/t/cannot-filter-whiskers/5279 made me realize that there's a really unfortunate distinction. While glyphs don't make a distinction between single/multiple items, annotations do. I don't know the reason behind this but it seems like something that should definitely be considered for removal. The other thing is that annotations by default don't support views for some reason - I guess nobody has requested it before, or nobody has got their hands on it.

**Comment by bryevdv** [comment]
I guess I would say yes and no, and it depends on the annotation. Box annotations are not usually attached to other glyphs. A plot may typically only have one or two box annotations, for example. The *entire value* in having a separate API at all from rect, say, is to make them simple to work with as individual things. 

But other cases the annotations *are* often "attached" to glyphs. Whiskers and Labels are probably the two most obvious examples. Perhaps we could embrace this taxonomy and recognize annotations that "attach" to other glyphs, one ones that are more typically for an entire plot (e.g. boxes, regions). The ones that attach to glyphs would:

* be vectorized, with a source
* respect views
* have a convenience API to express the "attachment"

**Comment by jbednar** [comment]
Where the attachment is optional?  E.g. sometimes we use labels simply like scatterplots, where the label stands in for the point itself (e.g. when using a unicode symbol, not a word, or when making a word cloud or similar plot).  So it's great to automate the attachment, but I'd hope it was optional.

**Comment by bryevdv** [comment]
> Where the attachment is optional? 

> and PlotAnnotation would be singular (a box, a, region, ***a free-standing label***).

I think it's reasonable to have two types for glyph labels and plot labels. I just don't like the current names "Label" and "LableSet" at all.

**Comment by Pierretherenard** [comment]
I came here from this link: https://stackoverflow.com/questions/49848966/how-to-add-permanent-name-labels-not-interactive-ones-on-nodes-for-a-networkx/49946144?noredirect=1#comment108699588_49946144
I am using networkx 2.3, bokeh 1.4 and python 3.5. My goal lately is to use a circle to group some dots for display.
![Screen Shot 2020-04-28 at 11 51 00 AM](https://user-images.githubusercontent.com/23002446/80525841-b5db8500-8946-11ea-90d4-dc161704cd7e.png)
This x, y = zip(*graph_renderer.layout_provider.graph_layout.values())
will give me that:
-- 116 -0.48802644690383895 -0.8782336181528193 
-- 117 -0.11818255864550706 -0.5910243156833965 
And I will be able to display the label number of the node

But I have no way to go back to the @index or to the G.nodes.keys() from the xy position without using javascript or send the xy to a storage and fill the network with the position. There should be a better way to link the position of the data dot to the data node. The node index/key is the only link to the information from the text.

**Comment by bryevdv** [comment]
@Pierretherenard As I mentioned in the thread, it's possible that the layout is computed in the browser, and therefore there is no way in general to make the layout indices available on the Python side under all circumstances. So I'm more interested to find ways to express label "attachment" to glyphs, and considering your use case above, ways to allow conditioning label visibility on other factors (e.g. selection) I think that would satisfy your needs?

**Comment by Pierretherenard** [comment]
conditional visibility yes if It means I can pass a list to control which glyph will display a label

**Comment by jeremy9959** [comment]
I am interested in this issue, in particular as it pertains to drawing graphs. When I drew a labelled graph I computed the node positions in python and added a label set using those positions.  @bryevdv mentioned this in the stackoverflow post referenced above:

>I should add, there are certainly other ways to do this. Since nx graphs generate static layouts, you >could also extract the (x,y) coordinates in Python and put them in a data source for the LabelSet. >That would avoid the JS transform, but I would regard the approach above as the most general and >"correct". 

Why is the JS approach "the most general and correct"?

**Comment by jbednar** [comment]
The JS approach would presumably leave the labels attached to the glyphs that they are labelling. In a static plot you won't see any difference, but what if you drag a node to a new position using one of the drawing tools?  If the label coordinates are computed only once, in Python, the label will stay where it was.

**Comment by bryevdv** [comment]
Yes, the best thing to do is have a first-class way to represent the *association itself*. That way labeling can be robust against layout updates regardless of where or how they originate.

**Comment by jeremy9959** [comment]
OK thanks now I understand the issue better.

**Comment by chrisbarber** [comment]
> https://discourse.bokeh.org/t/cannot-filter-whiskers/5279 made me realize that there's a really unfortunate distinction. While glyphs don't make a distinction between single/multiple items, annotations do. I don't know the reason behind this but it seems like something that should definitely be considered for removal. The other thing is that annotations by default don't support views for some reason - I guess nobody has requested it before, or nobody has got their hands on it.

+1 on supporting views. I just tried to do so with `LabelSet`. Also looking for an alternative for the meantime. It looked from the code example in this issue https://github.com/bokeh/bokeh/issues/6820 that one could supply a view to the `Text` glyph. But now I guess that ability has been removed? Just need some way of displaying text that vectorizes from a data source and also respects a view. Previously I was doing manual `Label`'s, but the code is so much cleaner being able to use a datasource like this.

**Comment by chrisbarber** [comment]
My bad; I just noticed that the view parameter goes to `add_glyph`, not to `Text`. I think it's working now.

**Comment by bryevdv** [comment]
I think this discussion can be closed now. The current plan is to move everything to be a glyph, more or less, on the same footing with respect to features, i.e. instead of having annotations be distinct "things" in the object system, have annotation be something a user _does_ (with whatever glyph is appropriate)
